### ISSUE TITLE
Package resource should not require users to keep multiple array indices in sync when installing multiple explicitly versioned packages

### ISSUE THREAD

**Issue by charlesjohnson** [chef/chef#3040]
As first raised by @mfischer-zd on the mailing list:

Chef 12.1 introduces the ability to install multiple packages simultaneously via the package resource, using the following syntax:

When using the new multi-package installation feature of the package resource introduced in Chef 12.1, multiple packages can be installed to yum & apt providers at once via this syntax:

```
package ['package1', 'package2']
```

However, to specify the versions of these packages, an array of versions must also be supplied, like so:

```
package ["foo", "bar"] do
  version ["1.0.0", "2.1.7"]
end
```

This implementation requires that users keep the indices of multiple arrays in sync by hand, which is a hostile interface. A single off-by-one error on either array index could have catastrophic unintended consequences, changing the installed versions of many packages simultaneously.

One proposed alternate implementation that may fix this: Instead, when installing multiple packages, the version field should be required to be a hash of package names and versions, and the package_name attribute should be ignored.

Example:

```
package 'package_batch' do
  version { "foo" => "1.0.0", "bar", => "2.1.7" }
end
```

One flaw I see in this approach that could lead to unexpected behavior, occurs when an array of packages is provided in the package_name attribute, while a hash is supplied in the version attribute. For instance:

```
package ['package1', 'package2'] do
  version { "foo" => "1.0.0", "bar", => "2.1.7", "baz", => "1.21.0" }
end
```

To deal with the case when an array is detected in the package_name attribute and a hash is detected in the version attribute, chef should return a warning explaining that the package_name attribute has been ignored because a hash was detected in the version attribute.

**Comment by mfischer-zd** [comment]
Another issue with the new syntax is that it introduces the concept of something other than a String as a resource name, which makes me wonder how the resource can be notified.  (Perhaps this isn't a big deal with packages, as I'm not sure you'd want to notify them, but the pattern is novel and undocumented.)  And it has changed to allow an Array as a name, but not a Hash.

Given all the ambiguities in behavior this can cause, I'm wondering if a new resource class should be created to handle multi-package installations instead of trying to jury-rig it into the legacy package resource class.

**Comment by lamont-granquist** [comment]
Notifies work like this:

```
package [ "foo", "bar" ] do
  action :nothing
end

log "notify it" do
  notifies :install, "package[foo, bar]", :immediately
end
```

That has been documented in chef-docs: https://docs.chef.io/resource_package.html#multiple-packages

We spent some time ensuring that Objects injected as names into resources were coerced into strings deterministically now:

https://github.com/chef/chef/blob/master/lib/chef/resource.rb#L122-L154

That also correctly fixes the use case that some people were using of passing Exception objects directly into the log resource.

The multipackage cookbook that I wrote:

https://github.com/lamont-cookbooks/multipackage

Also works around the title issue:

https://github.com/lamont-cookbooks/multipackage/blob/master/definitions/default.rb#L31-L32

So this works:

```
multipackage_install "ntp"
multipackage_install "strace"
multipackage_install "foo" do
  version "1.2.3"
end
```

becomes:

```
package [ "ntp", "strace", "foo" ] do
  version [ nil, nil, "1.2.3" ]
end
```

**Comment by charlesjohnson** [comment]
The multipackage behavior is compelling, and it might also lend itself to a package_batch block, similar to the way that the machine_batch blocks work in Provisioning.

However, it doesn't solve for the lack of safety in the package resource with version arrays. I still feel pretty strongly that asking users to sync multiple array indices by hand is too unsafe to leave as-is.

**Comment by mfischer-zd** [comment]
As a user, I do not want to require a cookbook and third-party resource provider in order to work around a basic resource definition shortcoming that should have been addressed and RFCed before released.

**Comment by mfischer-zd** [comment]
Also, the name-referencing hack is pretty bad, as it's whitespace-dependent.

**Comment by danielsdeleo** [comment]
FWIW, you can just do:

```ruby
pkg_map = { "foo" => "1.0.0", "bar", => "2.1.7", "baz", => "1.21.0" }
pkg_names = pkg_map.keys
pkg_versions = pkg_map.values
```

**Comment by lamont-granquist** [comment]
If anyone ever tries to fix this a complication that needs to be addressed is this:

```ruby
yum_package [ "foo", "bar" ] do
  version [ "1.0.0", "2.1.7" ]
  arch [ "i686", "x86_64" ]
end
```

The name isn't necessarily unique because the yum package provider introduces the arch array which is indexed the same as the name and arch arrays.

This doesn't work:

```
{
  foo: {
    version: "1.0.0",
    arch: "i686",
  },
  foo: {
    version: "2.0.0",
    arch: "x86_64",
  }
}
```

this could work:

{
   "foo.i686" => "1.0.0",
   "foo.x86_64" => "1.0.0",
}

but you have to take into account the fact that the arch array is optional.

gets even worse when you really dig into it and you discover that whatprovides syntax in the yum provider is supported so "names" may look like "foo > 1.29" or even just filenames like "/usr/bin/perl" so the simple concatenation there of name + arch will fail in edge cases.  it seems simple when consider package names and versions to be simple identifiers, but it isn't.

this is giving me a massive headache for entirely unrelated reasons where i've been making some implicit assumptions that "package_name" can be uniquely and sanely constructed for each array entry -- and i'm settling down on the idea that it much more complicated than that, and potentially will become more complicated in the future -- so that the unique identifier needs to remain the index on the arrays.

so TL;DR i'm never fixing this, and i doubt anyone else is going to patch it to adequate satisfaction to take into account all the stupid edge cases in how complicated we've made the package providers.  the resultant code would also be difficult to maintain and debug.

the multipackage cookbook is now in chef-cookbooks, maintained by the cookbooks team, used in other community cookbooks, and could ultimately get pulled into core which is the right way to address this problem.

anyone who disagrees with me should submit a PR and show me how i'm wrong rather than trying to argue with me about why i'm wrong -- i've fought with the code for the past 18 months and i'm pretty certain words won't change my mind.
